misc/runbench.py

```diff
@@ -203,9 +238,113 @@ def make_pylibmc_client(servers, **kw):
     return Prefix(__import__('pylibmc').Client(servers_, **kw), prefix)
 
 
+class Pool:
+    ''' adapted from pylibmc '''
+
+    client = libmc.ClientUnsafe
+
+    def __init__(self, *args, **kwargs):
+        self.args, self.kwargs = args, kwargs
+
+    def clone(self):
+        return self.client(*self.args, **self.kwargs)
+
+    def __getattr__(self, key):
+        if not hasattr(libmc.Client, key):
+            raise AttributeError
+        result = getattr(libmc.Client, key)
+        if callable(result):
+            @wraps(result)
+            def wrapper(*args, **kwargs):
+                with self.reserve() as mc:
+                    return getattr(mc, key)(*args, **kwargs)
+            return wrapper
+        return result
+
+
+class ThreadMappedPool(Pool):
+    client = libmc.Client
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.clients = {}
+
+    @property
+    def current_key(self):
+        return threading.current_thread().native_id
+
+    @contextmanager
+    def reserve(self):
+        key = self.current_key
+        mc = self.clients.pop(key, None)
+        if mc is None:
+            mc = self.clone()
+        try:
+            yield mc
+        finally:
+            self.clients[key] = mc
+
+
+class ThreadPool(Pool):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.clients = Queue()
+        for _ in range(POOL_SIZE):
+            self.clients.put(self.clone())
+
+    @contextmanager
+    def reserve(self):
+        mc = self.clients.get()
+        try:
+            yield mc
+        finally:
+            self.clients.put(mc)
+
+
+class BenchmarkThreadedClient(libmc.ThreadedClient):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.config(libmc.MC_INITIAL_CLIENTS, POOL_SIZE)
+        self.config(libmc.MC_MAX_CLIENTS, POOL_SIZE)
+
+
+class FIFOThreadPool(ThreadPool):
```

Comment:
This class is analogous to the C++ `LockPool` class and might be an easier way to understand the overall control flow